### ISSUE TITLE
symlinks are copied, not preserved.

### ISSUE THREAD

**Issue by dan-da** [maidsafe/safe-api#444]
I would expect symlinks to be preserved by default with an option to make a copy.  Behavior should be configurable based on whether symlink target is within the set of PUT files or else points to an outside location.   Would be useful to look at what scp, rsync, tar, zip, git, etc do with symlinks.

See also #410.

```
$ tree testdata/
testdata/
├── another.md
├── emptyfolder
├── noextension
├── subfolder
│   ├── sub2.md
│   └── subexists.md
├── testlink.md -> test.md
└── test.md
```

```
$ safe files put --recursive testdata
FilesContainer created at: "safe://hnyynys9wzifqez6w8z75keoeexcnhb7ppirqogsinenx55uqpxfa734qebnc"
+  testdata/another.md              safe://hbhynydt5b95dmumcm8yig4u1keuuh8hgsr5yx39xn4mqikp91sbdhbpwp 
+  testdata/noextension             safe://hbyyyynw7fi7ko8u3wtai6eekko3m5y6u9ud7w1i4ajycmwfwu3wwpnn81 
+  testdata/subfolder/sub2.md       safe://hbhynydyr41tz98funce1eqzkgju8e1yb4343ppct4fiwie5n3qhsw7o7x 
+  testdata/subfolder/subexists.md  safe://hbhynyn9uodh1ju5uzyti3gmmtwburrssd89rcwcy3rzofdpypwomrzzte 
+  testdata/test.md                 safe://hbhynydpan7d94mwp1bun3mxfnrfrui131an7ihu11wsn8dkr8odab9qwn 
+  testdata/testlink.md             safe://hbhynydpan7d94mwp1bun3mxfnrfrui131an7ihu11wsn8dkr8odab9qwn
```

```
$ safe files tree safe://hnyynys9wzifqez6w8z75keoeexcnhb7ppirqogsinenx55uqpxfa734qebnc
safe://hnyynys9wzifqez6w8z75keoeexcnhb7ppirqogsinenx55uqpxfa734qebnc
└── testdata
    ├── another.md
    ├── noextension
    ├── subfolder
    │   ├── sub2.md
    │   └── subexists.md
    ├── test.md
    └── testlink.md
```

**Comment by theWebalyst** [comment]
This is a good thing to think through and try to get right. IIRC rsync has the option to copy or follow symlinks. Another option would be to ignore. 

For copy, we'd need to decide what information a symlink should preserve which might depend on what is being copied. 

For example, if the symlink points to something that is to be copied to the network, the symlink should point to that, but what if it points to something that is not being copied? Does it preserve an absolute path or a relative path?

The behaviours could get confusing when the data is restored to a different device or different location in the same device, or files have been moved in the mean time etc. So we might want to be cautious and not allow copying, or try to design the behaviour to be as understandable as possible. I agree it's a good idea to look at those other programs, particularly `tar` and `rsync`.

**Comment by dan-da** [comment]
This was addressed by #580.  Symlinks are now stored, retrieved, and resolvable unless --follow-links flag is passed to upload operations.  Closing.

**Comment by dan-da** [comment]
Symlinks on windows are very tricky.   Other tools have problems with them and can exhibit gotchas, such as git and cygwin.   ( official git behaves differently from cygwin git even ).  So here's a few notes, and links to more info.

The following assumes the windows user is running as administrator, in developer mode, or otherwise has permissions to write symlinks.

1. cygwin git is basically unusable for checking out the test_symlinks folder at this time. By default, cygwin git checks out symlinks as Junction Points when the target does not exist.  This is because it relies on cygwin lib behavior, and seemingly does not use git's own knowledge of the target's type (dir or file).   Somehow cygwin programs can use/interpret the Junction Points as symlinks, but the target does not show up in "Command Prompt" and gives an error for rust read_link() function, so causes safe-api test failure.   Possibly/probably it could be worked around by changing the naming inside the test_symlinks folder so that the targets are lower alphabetically than the symlinks, and thereby get created first.  Alternatively it must be possible to read the junction point targets somehow, so that could be looked into in the future.  Here some related info:

* https://stackoverflow.com/questions/3648819/how-to-make-symbolic-link-with-cygwin-in-windows-7#18660841
* https://cygwin.com/cygwin-ug-net/using-cygwinenv.html

2. The official git distribution can create the symlinks properly on NTFS systems, but does not do so by default.  Instead, the setting core.symlinks=true must be present.  eg:
* git clone --config core.symlinks=true  <git-url>       -- or --
* git config --global core.symlinks true && git clone <git-url>

Official Git, Cloning a repo with iinks on windows.  (Settings)
https://stackoverflow.com/questions/11662868/what-happens-when-i-clone-a-repository-with-symlinks-on-windows

3. Possible improvements to the safe-api impl of symlink support on windows could be:
* ability to read Junction Points as symlinks, ala cygwin
* ability to fallback to write/read symlinks as specially formatted shortcut files, ala cygwin when perms not available.
